src/jit/flowgraph.cpp

```diff
@@ -22471,3 +22471,240 @@ void Compiler::fgLclFldAssign(unsigned lclNum)
+
+        // If there are more callfinallys than continuations, some of the
+        // callfinallys must share a continuation, and we can merge them.
+        const bool tryMerge = callFinallyCount > continuationMap.GetCount();
+
+        if (!tryMerge)
+        {
+            JITDUMP("EH#%i does not have any mergeable callfinallys\n", XTnum);
+            continue;
+        }
+
+        canMerge = true;
+
+        // Walk the try and callfinally regions, looking for blocks
+        // that jump to a callfinally that invokes this try's finally,
+        // and make sure they all jump to the appropriate canonical
+        // callfinally.
+        BasicBlock* const endTryBlock = HBtab->ebdTryLast->bbNext;
+        for (BasicBlock* currentBlock = HBtab->ebdTryBeg; currentBlock != endTryBlock;
+             currentBlock             = currentBlock->bbNext)
+        {
+            bool merged = fgRetargetBranchesToCanonicalCallFinally(currentBlock, beginHandlerBlock, continuationMap);
+            didMerge    = didMerge || merged;
+        }
+
+        for (BasicBlock* currentBlock = firstCallFinallyRangeBlock; currentBlock != endCallFinallyRangeBlock;
+             currentBlock             = currentBlock->bbNext)
+        {
+            bool merged = fgRetargetBranchesToCanonicalCallFinally(currentBlock, beginHandlerBlock, continuationMap);
+            didMerge    = didMerge || merged;
+        }
+    }
+
+#if DEBUG
+
+    if (!canMerge)
+    {
+        JITDUMP("Method had try-finallys, but did not have any mergeable finally chains.\n");
+        return;
+    }
+
+    assert(didMerge);
+    JITDUMP("Method had try-finallys and some callfinally merges were performed.\n");
+
+    if (verbose)
+    {
+        printf("\n*************** After fgMergeFinallyChains()\n");
+        fgDispBasicBlocks();
+        fgDispHandlerTab();
+        printf("\n");
+    }
+
+#endif // DEBUG
+}
+
+//------------------------------------------------------------------------
+// fgRetargetBranchesToCanonicalCallFinally: find non-canonical callfinally
+// invocations and make them canonical.
+//
+// Arguments:
+//     block -- block to examine for call finally invocation
+//     handler -- start of the finally region for the try
+//     continuationMap -- map giving the canonical callfinally for
+//        each continuation
+//
+// Returns:
+//     true iff the block's branch was retargeted.
+
+bool Compiler::fgRetargetBranchesToCanonicalCallFinally(BasicBlock*      block,
+                                                        BasicBlock*      handler,
+                                                        BlockToBlockMap& continuationMap)
+{
+    // We expect callfinallys to be invoked by a BBJ_ALWAYS at this
+    // stage in compilation.
+    if (block->bbJumpKind != BBJ_ALWAYS)
+    {
+        // Possible paranoia assert here -- no flow successor of
+        // this block should be a callfinally for this try.
+        return false;
+    }
```

Comment:
So, on arm32, if you have:
```
try1 {
   try2 {
      callfinally F2 / always L1 // A
      ...
      callfinally F2 / always L2 // B
      ...
   }
   L1: callfinally F1 / always L3 // C
   L2: callfinally F1 / always L3 // D
}
L3:

finally2: ... // F2
finally1: ... // F1
```

So we want to merge L1 & L2. Say L1 is canonical. When processing the 'always' of 'B' you would change it point to L1. Interestingly, that seems like it should work, although I don't think that's what you were thinking the structure of the callfinally blocks was when you wrote the code. I wonder if there's some more complex case I'm missing.

On amd64 you would have seen:
```
try1 {
   try2 {
      always A1
      ...
      always A2
      ...
   }
   A1: callfinally F2 / always L1 // A
   A2: callfinally F2 / always L2 // B
}
L1: callfinally F1 / always L3 // C
L2: callfinally F1 / always L3 // D
L3:

finally2: ... // F2
finally1: ... // F1
```